File: backend/app/schemas/analytics.py
from typing import List

from pydantic import BaseModel


class ListingAnalyticsItem(BaseModel):
    id: int
    title: str

    views_count: int
    favorites_count: int
    messages_received: int
    availability_confirmations_count: int

    # Percentage of views that turned into a
    # tenant starting a conversation, 0-100.
    conversion_rate: float

    class Config:
        from_attributes = True


class LandlordAnalyticsResponse(BaseModel):
    total_views: int
    total_favorites: int
    total_messages_received: int
    total_availability_confirmations: int

    # Percentage of views that turned into a
    # tenant starting a conversation, 0-100.
    conversion_rate: float

    listings: List[ListingAnalyticsItem]


class ReportsBreakdown(BaseModel):
    pending: int
    dismissed: int
    action_taken: int


class AdminAnalyticsResponse(BaseModel):
    # Non-admin users who logged in within
    # the last 30 days.
    active_users: int

    listings_created: int
    pending_approvals: int
    suspended_listings: int

    reports_received: int
    reports_breakdown: ReportsBreakdown

File: frontend/src/types/analytics.ts
export type ListingAnalyticsItem = {
  id: number;
  title: string;

  views_count: number;
  favorites_count: number;
  messages_received: number;
  availability_confirmations_count: number;

  // Percentage, 0-100.
  conversion_rate: number;
};

export type LandlordAnalyticsResponse = {
  total_views: number;
  total_favorites: number;
  total_messages_received: number;
  total_availability_confirmations: number;

  // Percentage, 0-100.
  conversion_rate: number;

  listings: ListingAnalyticsItem[];
};

export type ReportsBreakdown = {
  pending: number;
  dismissed: number;
  action_taken: number;
};

export type AdminAnalyticsResponse = {
  active_users: number;
  listings_created: number;
  pending_approvals: number;
  suspended_listings: number;

  reports_received: number;
  reports_breakdown: ReportsBreakdown;
};
